### PR TITLE
Pickup Tests

### PR DESCRIPTION
**Pull request by OlafSzmidt**
Initial dig at writing more extensive unit pickup tests. Coverage of the file was fine but they clearly weren't tested well enough. I'm new to tests so criticize as much as you can :P

- `HealthPickup` has now certain constraints. It can't be equal to or less than 0, and can't be greater than 100.
- If the pickup added to the avatar health will be greater than 100, cap it at 100hp.
- DamagePickup can be now specified with the custom damage boost.
- Pickups are more testable. You can make use of `set_custom_pickup` to add your own params as opposed to using only default ones.
- Extra tests.

Please test extensively before approving. We **don't** have integration tests anymore. Use the manual test plan to get an idea of things that need to be checked.

### CODE REVIEW

File: aimmo-game/simulation/pickups.py [hunk 1]
```diff
@@ -44,6 +51,10 @@ def serialise(self):
     def _apply(self, avatar):
         avatar.health += self.health_restored
 
+        # Make sure the health is capped at 100.
```
Comment (CelineBoudier) [inline aimmo-game/simulation/pickups.py]: Isn't that max number bound to change with say, achievements / points...? Ie can we have this max number not hard coded? cc @Spycho

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 2]
```diff
@@ -44,18 +48,72 @@ def test_effect_added(self):
...
+        self.assertEqual(self.avatar.health, 15)
+
+    def test_health_increases_custom_positive_50(self):
+        self.avatar.health = 10
+        self.set_custom_pickup(50)
+        self.apply_pickup()
+        self.assertEqual(self.avatar.health, 60)
+
+    # Checking that the health gain argument raises exceptions in correct scenarios.
+    def test_health_argument_raises_exception_one(self):
```
Comment (CelineBoudier) [inline aimmo-game/tests/test_simulation/test_pickups.py]: no descriptive enough - please replace one two and three with a short description of what the scenario is

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 3]
```diff
@@ -44,18 +48,72 @@ def test_effect_added(self):
...
+        self.assertRaises(ValueError, self.set_custom_pickup, 150)
+
+    def test_health_argument_raises_exception_two(self):
+        self.assertRaises(ValueError, self.set_custom_pickup, 0)
+
+    def test_health_argument_raises_exception_three(self):
+        self.assertRaises(ValueError, self.set_custom_pickup, -50)
+
+    def test_health_argument_does_not_raise_exception_one(self):
```
Comment (CelineBoudier) [inline aimmo-game/tests/test_simulation/test_pickups.py]: same

File: aimmo-game/simulation/pickups.py (hunk 1 above)
Comment (duncanrussell-ocado) [reply aimmo-game/simulation/pickups.py]: This is a wider question about how we process turns. It is possible for an avatar (on 50 health) in a single turn to get both 100 points health pickup and -80 points damage pickup?
Is the result 70 points (50+100-80), or is the result 20 points (50+100, limited to 100, - 80)?

File: aimmo-game/tests/test_simulation/test_pickups.py (hunk 2 above)
Comment (duncanrussell-ocado) [reply aimmo-game/tests/test_simulation/test_pickups.py]: ... or reduce to one test the for same responsibility.

File: aimmo-game/simulation/pickups.py (hunk 1 above)
Comment (OlafSzmidt) [reply aimmo-game/simulation/pickups.py]: @CelineBoudier: I wasn't aware that this is what we're planning to do, so I can remove this cap or we can change it when we have the extra functionality of achievements or points in our game.

@duncanrussell-ocado: So there's a few things that you're misunderstanding about the game logic I think. Damage pickups don't actually damage the avatar, they give them a damage boost if they want to attack others (this is a near refactoring goal for sure). Another point is, one cell cannot hold two pickups, so the scenario isn't possible.

File: aimmo-game/simulation/pickups.py [hunk 4]
```diff
@@ -29,8 +29,15 @@ def serialise(self):
 
 class HealthPickup(_Pickup):
     def __init__(self, cell, health_restored=3):
```
[the remark this answers is not in the document]
Comment (OlafSzmidt) [reply aimmo-game/simulation/pickups.py]: It's just a default value for the pickup if health_restored isn't specified. It can be changed but it's all down to discussion as to how much goes where tbh.

File: aimmo-game/simulation/pickups.py (hunk 1 above)
Comment (OlafSzmidt) [reply aimmo-game/simulation/pickups.py]: So as discussed with Chris, this is actually a wider topic we need to cover. I'll make an issue about this. For now, we want any kind of capping (whether it's 10 or 100, it makes no difference for now) functionality because limitless HP isn't something we generally want? :)

File: aimmo-game/simulation/pickups.py [hunk 5]
```diff
@@ -29,8 +29,15 @@ def serialise(self):
...
-        super(HealthPickup, self).__init__(cell)
-        self.health_restored = health_restored
+        # Round the value to the nearest integer.
+        health_restored = int(round(health_restored))
+
+        # Check if the value provided is legal.
+        if 0 < health_restored <= 100:
```
[the remark this answers is not in the document]
Comment (OlafSzmidt) [reply aimmo-game/simulation/pickups.py]: It's actually not. @Spycho believes it's 10, but there's not capping logic in the code right now (I just tested it, currently my avatar is at 68). Will look into this and make changes.

File: aimmo-game/simulation/pickups.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dips7189) [reply aimmo-game/simulation/pickups.py]: ok.

File: aimmo-game/simulation/pickups.py [hunk 6]
```diff
@@ -29,8 +29,15 @@ def serialise(self):
 
 class HealthPickup(_Pickup):
     def __init__(self, cell, health_restored=3):
-        super(HealthPickup, self).__init__(cell)
-        self.health_restored = health_restored
+        # Round the value to the nearest integer.
+        health_restored = int(round(health_restored))
```
[the remark this answers is not in the document]
Comment (OlafSzmidt) [reply aimmo-game/simulation/pickups.py]: So as we discussed yesterday, the user does not access this method directly so checking is not required (?). A 'generic' exception is thrown by round() and that's adequate enough.

I see where you are coming from, because we try to think like Java, but don't think it's a pythonic way of checking types for every method param.

File: aimmo-game/simulation/pickups.py [hunk 7]
```diff
@@ -73,9 +84,9 @@ def serialise(self):
 class DamagePickup(_PickupEffect):
```
[the remark this answers is not in the document]
Comment (OlafSzmidt) [reply aimmo-game/simulation/pickups.py]: :+1:

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 8]
```diff
@@ -44,18 +48,72 @@ def test_effect_added(self):
...
 
+    def test_invulnerability_is_applied(self):
+        self.apply_pickup()
+        self.assertIsInstance(self.pickup, pickups.HealthPickup)
+
     def test_health_increases(self):
         self.apply_pickup()
         self.assertEqual(self.avatar.health, 8)
 
     def test_serialise(self):
```
[the remark this answers is not in the document]
Comment (OlafSzmidt) [reply aimmo-game/tests/test_simulation/test_pickups.py]: Asked @Spycho about this :)

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 9]
```diff
@@ -44,18 +48,72 @@ def test_effect_added(self):
...
 
+    # Test that the health pickup with a custom health gain value is applied correctly. We do not check
+    # for restraints (ie. less than 0 or more than 100) in the following tests:
+    def test_health_increases_custom_positive_5(self):
+        self.avatar.health = 10
+        self.set_custom_pickup(5)
+        self.apply_pickup()
+        self.assertEqual(self.avatar.health, 15)
+
+    def test_health_increases_custom_positive_50(self):
```
[the remark this answers is not in the document]
Comment (OlafSzmidt) [reply aimmo-game/tests/test_simulation/test_pickups.py]: Yes I have moved them into the same method now.

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 10]
```diff
@@ -44,18 +48,72 @@ def test_effect_added(self):
...
+        except ValueError:
+            self.fail("set_custom_pickup() raised ValueError unexpectedly!")
+
+    def test_health_cannot_be_greater_than_100(self):
+        self.avatar.health = 90
+        self.set_custom_pickup(20)
+        self.apply_pickup()
+        self.assertEqual(self.avatar.health, 100)
+
+    def test_health_rounding_to_nearest_int(self):
```
[the remark this answers is not in the document]
Comment (OlafSzmidt) [reply aimmo-game/tests/test_simulation/test_pickups.py]: We have agreed so now, yes.

File: aimmo-game/tests/test_simulation/test_pickups.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dips7189) [reply aimmo-game/tests/test_simulation/test_pickups.py]: so is ok to remove? Or ...

[the remark this answers is not in the document]
Comment (OlafSzmidt) [reply aimmo-game/tests/test_simulation/test_pickups.py]: Don't think so. It's not a web service that we're running but this is sitting on flask so don't think we have pickups as django objects for example.

Some files have serialise() methods in them exactly because of this. Not 100% on this.

[the remark this answers is not in the document]
Comment (dips7189) [reply aimmo-game/tests/test_simulation/test_pickups.py]: ok, that's cool

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 11]
```diff
@@ -44,25 +47,85 @@ def test_effect_added(self):
...
+        self.assertEqual(self.avatar.health, 10)
+
+        # Test 2 - increase of 50
+        self.avatar.health = 10
+        self.set_custom_pickup(50)
+        self.apply_pickup()
+        self.assertEqual(self.avatar.health, 60)
+
+    def test_health_argument_raises_exception(self):
+        self.assertRaises(ValueError, self.set_custom_pickup, 150)
```
Comment (dips7189) [inline aimmo-game/tests/test_simulation/test_pickups.py]: This test case is testing three edge cases.
Ideally we want to have three tests each testing the specific edge case, otherwise it is just hiding the fact that there are edge cases.

test_health_argument_raise_exception_when_health_exceeded
etc ...

These tests make it clear what the edge cases are instead of grouping them based on their return value or thrown value. That isn't how tests should be written IMO.

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 12]
```diff
@@ -44,25 +47,85 @@ def test_effect_added(self):
...
+        self.assertRaises(ValueError, self.set_custom_pickup, 0)
+        self.assertRaises(ValueError, self.set_custom_pickup, -50)
+
+    def test_health_argument_does_not_raise_exception(self):
+        try:
```
Comment (dips7189) [inline aimmo-game/tests/test_simulation/test_pickups.py]: as per discussion, test health increases (happy path test)

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 13]
```diff
@@ -44,25 +47,85 @@ def test_effect_added(self):
...
     def test_health_increases(self):
         self.apply_pickup()
         self.assertEqual(self.avatar.health, 8)
 
     def test_serialise(self):
         self.assertEqual(self.pickup.serialise(), {'type': 'health', 'health_restored': 3})
 
+    def test_given_custom_health_pickup_increase_health_increases(self):
+        # Test 1 - increase of 5
+        self.assertEqual(self.avatar.health, 5)
```
Comment (dips7189) [inline aimmo-game/tests/test_simulation/test_pickups.py]: consider the following style of test:

// given
self.avatar.health = 10
self.set_custom_pickup(5)

// when
self.apply_pickup()

// then
self.assertEqual(self.avatar.health, 15)

You don't have to put the comments but the separation makes it clear what is happening.

File: aimmo-game/simulation/pickups.py [hunk 14]
```diff
@@ -81,16 +80,16 @@ def serialise(self):
         }
 
 
-class DamagePickup(_PickupEffect):
+class DamageBoostPickup(_PickupEffect):
```
Comment (CelineBoudier) [inline aimmo-game/simulation/pickups.py]: that's fine but the refactoring could be another pull request as not really linked with test... :)

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 15]
```diff
@@ -59,45 +58,38 @@ def test_health_increases(self):
     def test_serialise(self):
         self.assertEqual(self.pickup.serialise(), {'type': 'health', 'health_restored': 3})
 
-    # Test that the health pickup with a custom health gain value is applied correctly. We do not check
-    # for restraints (ie. less than 0 or more than 100) in the following tests:
-    def test_health_increases_custom_positive_5(self):
-        self.avatar.health = 10
+    def test_given_custom_health_pickup_increase_health_increases(self):
+        # Test 1 - increase of 5
+        self.assertEqual(self.avatar.health, 5)
```
Comment (CelineBoudier) [inline aimmo-game/tests/test_simulation/test_pickups.py]: wait where is the avatar health set again?

File: aimmo-game/tests/test_simulation/test_pickups.py (hunk 11 above)
Comment (CelineBoudier) [reply aimmo-game/tests/test_simulation/test_pickups.py]: +1

File: aimmo-game/tests/test_simulation/test_pickups.py [hunk 16]
```diff
@@ -59,45 +58,38 @@ def test_health_increases(self):
...
         self.assertRaises(ValueError, self.set_custom_pickup, 150)
-
-    def test_health_argument_raises_exception_two(self):
         self.assertRaises(ValueError, self.set_custom_pickup, 0)
-
-    def test_health_argument_raises_exception_three(self):
         self.assertRaises(ValueError, self.set_custom_pickup, -50)
 
-    def test_health_argument_does_not_raise_exception_one(self):
+    def test_health_argument_does_not_raise_exception(self):
```
Comment (CelineBoudier) [inline aimmo-game/tests/test_simulation/test_pickups.py]: same as per @dips7189 comment above

File: aimmo-game/simulation/pickups.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dips7189) [reply aimmo-game/simulation/pickups.py]: kool

File: aimmo-game/simulation/pickups.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dips7189) [reply aimmo-game/simulation/pickups.py]: ok

File: aimmo-game/tests/test_simulation/test_pickups.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dips7189) [reply aimmo-game/tests/test_simulation/test_pickups.py]: ok

File: aimmo-game/simulation/pickups.py (hunk 14 above)
Comment (OlafSzmidt) [reply aimmo-game/simulation/pickups.py]: Do we not want to refactor code as we see it?

File: aimmo-game/tests/test_simulation/test_pickups.py (hunk 15 above)
Comment (OlafSzmidt) [reply aimmo-game/tests/test_simulation/test_pickups.py]: In the text fixture, setUp()

File: aimmo-game/tests/test_simulation/test_pickups.py (hunk 13 above)
Comment (OlafSzmidt) [reply aimmo-game/tests/test_simulation/test_pickups.py]: ok cool

File: aimmo-game/simulation/pickups.py (hunk 14 above)
Comment (CelineBoudier) [reply aimmo-game/simulation/pickups.py]: Yes, in another pull request :) Else it's just confusing as the current PR is actually doing 2 or 3 things instead of what is described, and then makes it harder to deal with later. See some interesting things here: https://ro-che.info/articles/2016-09-18-good-pull-requests and there: https://blog.codacy.com/why-youre-not-merging-your-pull-requests-and-what-to-do-about-it-12701c70ae5c

Comment (CelineBoudier) [reply aimmo-game/simulation/pickups.py]: I'd rather you remove this from this PR and prepare a small pull request for this refactoring

Comment (OlafSzmidt) [reply aimmo-game/simulation/pickups.py]: :+1: okk

**Comment by OlafSzmidt** [comment]
Please make sure you test the game with `./run.py -k` before approving. We don't have integration tests ;)

**Comment by CelineBoudier** [comment]
Oh reallllly? ;)

**Comment by OlafSzmidt** [comment]
Yeeeaaaaaaa, we will write them soon though ==)

**Comment by CelineBoudier** [comment]
Please squash the commits that make sense together and I might approve

**Comment by OlafSzmidt** [comment]
Done;]
